src/core/network/polling.ts

```diff
@@ -0,0 +1,61 @@
+import * as tt from '../../telegram-types'
+import AbortController from 'abort-controller'
+import ApiClient from './client'
+import d from 'debug'
+import { promisify } from 'util'
+const wait = promisify(setTimeout)
+const debug = d('telegraf:polling')
+
+export class Polling {
+  private readonly abortController = new AbortController()
+  private offset = 0
+  constructor(
+    private readonly telegram: ApiClient,
+    private readonly allowedUpdates: readonly tt.UpdateType[]
+  ) {}
+
+  async *[Symbol.asyncIterator](): AsyncGenerator<tt.Update[], void, void> {
+    debug('Starting long polling')
+    do {
+      try {
+        const updates = await this.telegram.callApi(
+          'getUpdates',
+          {
+            timeout: 50,
+            offset: this.offset,
+            allowed_updates: this.allowedUpdates,
+          },
+          this.abortController
+        )
+        const last = updates[updates.length - 1]
+        if (last !== undefined) {
+          this.offset = last.update_id + 1
+        }
+        yield updates
+      } catch (err) {
+        if (err.name === 'AbortError') return
+        if (err.code === 401 || err.code === 409) throw err
+        const retryAfter: number = err.parameters?.retry_after ?? 5
+        debug('Failed to fetch updates, retrying after %ds.', retryAfter, err)
+        await wait(retryAfter * 1000)
```

Comment:
I can see that you copied this from the old code, but why is it even that we only rethrow upon 401 and 409? Wouldn't it make more sense to rather check for 429 or for `retry_after` and throw in all other cases?

Reply:
What errors are we dealing with and how should we handle them?

- `AbortError` - `return`
- 401 Unauthorized - unrecoverable, `throw err`
- 409 Conflict - unrecoverable, `throw err`
- 429 Too Many Requests - if it ever happens, wait and retry
- 5xx - wait and retry
- `FetchError` - wait and retry, it's probably connectivity issue
- Did I forget anything?
- Unknown errors - `throw err` I guess?

Note that throwing here = crashing.

Reply:
I wish I could use `continue` here.

Reply:
> `getUpdates` never returns 429.
https://t.me/tdlibchat/5368

Reply:
> I wish I could use `continue` here.

Why is that not possible? :thinking:

Reply:
Regarding the errors, we should maybe just leave it like that and try to figure out what's best over time. I cannot come up with an ad hoc solution here, especially not when we don't have any official source that tells use which errors could be returned (do we?).

Reply:
>> I wish I could use `continue` here.
>
> Why is that not possible? :thinking: 

Bare `if (...) continue` wouldn't `await wait(...)`, I had to get more creative than that.

I decided to retry only on `FetchError` and `>= 500` error codes, people will open issues if it's too harsh, at least we know we aren't hiding any important errors.